### ISSUE TITLE
Unable to assign None to __hash__

### ISSUE THREAD

**Issue by Sxderp** [microsoft/pyright#9919]
Related issue: https://github.com/microsoft/pyright/issues/3223#issuecomment-1073007948

My linter (and the above comment) recommends assigning `None` to `__hash__` rather than relying on the implicit behavior of Python. Pyright complains about this. 

```python
class MyA:
    __hash__ = None
```

Produces the following in the playground

> Type "None" is not assignable to declared type "(self: Self@object) -> int"
  Type "None" is not assignable to type "(self: Self@object) -> int"  (reportAssignmentType)

If I do what's suggested here : https://github.com/microsoft/pyright/issues/5446#issuecomment-1628228484

```python
class MyA:
    __hash__: None = None
```

I get 

> "__hash__" overrides method of same name in class "object" with incompatible type "None"  (reportIncompatibleMethodOverride)

I think Pyright should allow .. something? Am I doing this wrong?

**Comment by erictraut** [comment]
Pyright is enforcing the type declaration for `object.__hash__` provided [here](https://github.com/python/typeshed/blob/cc8ca939c0477a49fcce0554fa1743bd5c656a11/stdlib/builtins.pyi#L121) in the typeshed stubs. As such, it's doing the right thing based on the type information it's given.

There have been some [recent discussions](https://discuss.python.org/t/hash-eq-and-lsp/68138) in the Python typing forum about modify the typeshed stub definition of `object.__hash__` and augmenting the typing spec to provide more specific guidance about how type checkers should handle it, but there hasn't been any convergence yet. Unless/until there is, I'm going to leave pyright's current behavior as is.

You can use a `# type: ignore` or `# pyright: ignore` comment if you'd like to suppress the error you're seeing here.

The challenge with `__hash__` is that the basic behavior of this magic method in Python doesn't follow the Liksov Substitution Principle (LSP), which is core to nominal (inherited) types.
